xds/src/test/java/com/linecorp/armeria/xds/SotwXdsStreamTest.java

```diff
@@ -0,0 +1,315 @@
+            assertThat(responseHandler.getResponses()).isEmpty();
+        }
+    }
+
+    @Test
+    void restart() throws Exception {
+        final SotwDiscoveryStub stub = SotwDiscoveryStub.ads(GrpcClients.builder(server.httpUri()));
+        final SubscriberStorage subscriberStorage = new SubscriberStorage();
+        final TestResponseHandler responseHandler = new TestResponseHandler(subscriberStorage);
+
+        try (SotwXdsStream stream = new SotwXdsStream(
+                stub, SERVER_INFO, Backoff.ofDefault(), eventLoop.get(), responseHandler,
+                new SubscriberStorage())) {
+
+            Thread.sleep(100);
+            assertThat(responseHandler.getResponses()).isEmpty();
+
+            subscriberStorage.register(XdsType.CLUSTER, clusterName);
+            stream.start();
+
+            // check if the initial cache update is done
+            await().until(() -> !responseHandler.getResponses().isEmpty());
+            assertThat(responseHandler.getResponses()).allSatisfy(res -> {
+                final Cluster expected = cache.getSnapshot(GROUP).clusters().resources().get(clusterName);
+                assertThat(res.getResources(0).unpack(Cluster.class)).isEqualTo(expected);
+            });
+            responseHandler.clear();
+
+            // stop the stream and verify there are no updates
+            stream.stop();
+            await().until(() -> stream.requestObserver == null);
+
+            cache.setSnapshot(
+                    GROUP,
+                    Snapshot.create(
+                            ImmutableList.of(createCluster(clusterName, 1)),
+                            ImmutableList.of(), ImmutableList.of(), ImmutableList.of(),
+                            ImmutableList.of(), "2"));
+            Thread.sleep(100);
+            assertThat(responseHandler.getResponses()).isEmpty();
+
+            // restart the thread and verify that the handle receives the update
+            stream.start();
+            await().until(() -> !responseHandler.getResponses().isEmpty());
+            assertThat(responseHandler.getResponses()).allSatisfy(res -> {
+                final Cluster expected = cache.getSnapshot(GROUP).clusters().resources().get(clusterName);
+                assertThat(res.getResources(0).unpack(Cluster.class)).isEqualTo(expected);
+            });
+        }
+    }
+
+    @Test
+    void errorHandling() throws Exception {
+        final SotwDiscoveryStub stub = SotwDiscoveryStub.ads(GrpcClients.builder(server.httpUri()));
+        final SubscriberStorage subscriberStorage = new SubscriberStorage();
+        final AtomicLong cntRef = new AtomicLong();
+        final TestResponseHandler responseHandler = new TestResponseHandler(subscriberStorage) {
+            @Override
+            public <T extends Message> void handleResponse(
+                    ResourceParser<T> type, DiscoveryResponse value, SotwXdsStream sender) {
+                if (cntRef.getAndIncrement() == 0) {
```

Comment:
Shouldn't we use the backoff to make sure that the exception is raised?

Reply:
I'm not sure I understood 😅  Do you mean you would prefer that backoff is done more than once?